grapher/controls/EntitySelectorModal.tsx

```diff
@@ -11,62 +11,112 @@ interface SearchableEntity {
+                </li>
+            )
+        } else {
+            return (
+                <li
+                    className="clickable"
+                    onClick={(): void => onSelect(result.name)}
+                >
+                    {result.name}
+                </li>
+            )
+        }
+    }
+}
+
 @observer
-class EntitySelectorMulti extends React.Component<{
+export class EntitySelectorModal extends React.Component<{
     selectionArray: SelectionArray
+    isMulti: boolean
     onDismiss: () => void
 }> {
-    @observable searchInput?: string
+    @observable searchInput: string = ""
     searchField!: HTMLInputElement
     base: React.RefObject<HTMLDivElement> = React.createRef()
-    dismissable: boolean = true
 
-    @computed get availableEntities(): string[] {
-        return this.props.selectionArray.availableEntityNames
+    @computed get sortedAvailableEntities(): string[] {
+        return sortBy(this.props.selectionArray.availableEntityNames)
+    }
+
+    @computed get isMulti(): boolean {
+        return this.props.isMulti
     }
 
     @computed get fuzzy(): FuzzySearch<SearchableEntity> {
         return new FuzzySearch(this.searchableEntities, "name")
     }
 
     @computed private get searchableEntities(): SearchableEntity[] {
-        return this.availableEntities.map((name) => {
+        return this.sortedAvailableEntities.map((name) => {
             return { name } as SearchableEntity
         })
     }
 
     @computed get searchResults(): SearchableEntity[] {
         return this.searchInput
             ? this.fuzzy.search(this.searchInput)
-            : sortBy(this.searchableEntities, (result): any => result.name)
+            : this.searchableEntities
     }
 
-    @action.bound onClickOutside(e: MouseEvent): void {
-        if (this.dismissable) this.props.onDismiss()
+    @action.bound onSelect(entityName: string): void {
+        if (this.isMulti) {
+            this.props.selectionArray.toggleSelection(entityName)
+        } else {
+            this.props.selectionArray.setSelectedEntities([entityName])
+            this.props.onDismiss()
+        }
     }
 
-    componentDidMount(): void {
-        // HACK (Mispy): The normal ways of doing this (stopPropagation etc) don't seem to work here
-        this.base.current!.addEventListener("click", () => {
-            this.dismissable = false
-            setTimeout(() => (this.dismissable = true), 100)
-        })
-        setTimeout(
-            () => document.addEventListener("click", this.onClickOutside),
-            1
+    @action.bound onDocumentClick(e: MouseEvent): void {
+        // check if the click was outside of the modal
+        if (
+            this.base?.current &&
+            !this.base.current.contains(e.target as Node) &&
+            // check that the target is still mounted to the document; we also get click events on nodes that have since been removed by React
+            document.contains(e.target as Node)
```

Comment:
This here is what the old `HACK(Mispy)` was all about - we would get an event for one of the checkboxes on the right of the `EntitySelectorMulti`, but by the time the event handler triggers has already unmounted the (now-unchecked) checkbox.
And then the modal would close, because `this.base.current.contains(e.target)` is false.